### ISSUE TITLE
Enable Hardware address decoder

### ISSUE THREAD

**Issue by mariusvanhorssen** [n0mjs710/MCP23S17#6]
The way the hardware address decoder is enabled is incorrect.

- The IOCON register exists of 2 bytes, while one byte is written
- There is a silicon bug when hardware addressing is enabled: [http://ww1.microchip.com/downloads/en/DeviceDoc/80311a.pdf](url)

**Comment by n0mjs710** [comment]
Don't really care, it seems to work. BTW: I'm not charging you for this, so if you don't like the way it works, don't use it.

**Comment by mariusvanhorssen** [comment]
No, it only works for address 0 - 3. You need to modify the code to make it working for address 4 - 7.

**Comment by n0mjs710** [comment]
Both ICON registers are the same. One register, two addresses. Setting one sets both (data sheet section 3.5, first paragraph):

> One register (IOCON) is shared between the two ports. The PORTA registers are identical to the PORTB registers

The Errata indicates the problem exists when hardware addressing is DISABLED (emphasis added):

> 1. Module: SPI Module
> **When IOCON.HAEN = 0** (hardware addressing dis- abled): If the A2 pin is high, then the device must be addressed as A2, A1, A0, = 1xx (i.e., OPCODE = b”0100 1XX’.

When the function MCP::begin() is called, SPI bus is started and IOCON.HAEN is immediately set to "1", which ENABLES hardware addressing:

    ADDR_ENABLE   (0b00001000)
    byteWrite(IOCON, ADDR_ENABLE);

I currently use this class with addresses higher than 4 without issue.

**Comment by mariusvanhorssen** [comment]
> When the function MCP::begin() is called, SPI bus is started and IOCON.HAEN is immediately set to "1", which ENABLES hardware addressing.

This is were it goes wrong. When your constructor is called, addressing is disabled. So, when a chip is connected with pin A2 high, the addressing mode will not be enabled because the function byteWrite sends an invalid opcode.

**Comment by n0mjs710** [comment]
I'm not sure if it's the parts you have or what. I've been using this class in my own projects for years, I've never had an addressing problem with my parts, and I've used addresses >=4. I didn't just write this and walk away. I wrote it because I use it... extensively. I decided to post it on GitHub because I thought it might help others.

I've closed the issue. I'm not updating the class to fix a problem I cannot replicate. My advice is to fork the class and make your own local changes.
